google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,356 @@
+
+  /**
+   * The number of pending BulkWriter operations. Used to determine when the next query can be run.
+   */
+  private int pendingOperationsCount = 0;
+
+  RecursiveDelete(
+      FirestoreRpcContext<?> firestoreRpcContext,
+      BulkWriter writer,
+      CollectionReference reference) {
+    Preconditions.checkState(reference != null, "CollectionReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+
+    this.documentReference = null;
+
+    // The parent is the closest ancestor document to the location we're deleting. Since we are
+    // deleting a collection, the parent is the path of the document containing that collection (or
+    // the database root, if it is a root collection).
+    this.parentPath = reference.getResourcePath().popLast();
+    this.collectionId = reference.getId();
+  }
+
+  RecursiveDelete(
+      FirestoreImpl firestoreRpcContext, BulkWriter writer, DocumentReference reference) {
+    Preconditions.checkState(reference != null, "DocumentReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+
+    this.documentReference = reference;
+
+    // The parent is the closest ancestor document to the location we're deleting. Since we are
+    // deleting a document, the parent is the path of that document.
+    this.parentPath = documentReference.getResourcePath();
+    this.collectionId = documentReference.getParent().getId();
+  }
+
+  @VisibleForTesting
+  RecursiveDelete(
+      FirestoreRpcContext<?> firestoreRpcContext,
+      BulkWriter writer,
+      CollectionReference reference,
+      int maxLimit,
+      int minLimit) {
+    this(firestoreRpcContext, writer, reference);
+    this.maxPendingOps = maxLimit;
+    this.minPendingOps = minLimit;
+  }
+
+  public ApiFuture<Void> run() {
+    Preconditions.checkState(!invoked, "RecursiveDelete.run() should only be called once");
+    invoked = true;
+
+    writer.verifyNotClosed();
+
+    streamDescendants();
+    return completionFuture;
+  }
+
+  private void streamDescendants() {
+
+    Query query = getAllDescendantsQuery();
+    streamInProgress = true;
+    final int[] streamedDocsCount = {0};
+    final ApiStreamObserver<DocumentSnapshot> responseObserver =
+        new ApiStreamObserver<DocumentSnapshot>() {
+          public void onNext(DocumentSnapshot snapshot) {
+            streamedDocsCount[0]++;
+            lastDocumentSnap = snapshot;
+            deleteReference(snapshot.getReference());
+          }
+
+          public void onError(Throwable throwable) {
+            synchronized (lock) {
+              String message = "Failed to fetch children documents. " + throwable.getMessage();
+              lastError = new FirestoreException(message, Status.UNAVAILABLE);
+              onQueryEnd();
+            }
+          }
+
+          public void onCompleted() {
+            synchronized (lock) {
+              streamInProgress = false;
```

Comment:
We should probably reduce the scope of the lock to just this line (also for the onError callback, which only needs to set lastError under lock). This improves readability as it makes the intention more clear and also answer the question why `onNext` does not use locks.

Reply:
Done.